cxr-foundation/cxr_foundation/embeddings_data.py

```diff
@@ -0,0 +1,167 @@
+
+  """
+  raw_dataset = tf.data.TFRecordDataset(filename)
+  # Expect only one element in the TFRecord.
+  for raw_record in raw_dataset.take(1):
+    example = tf.train.Example()
+    example.ParseFromString(raw_record.numpy())
+
+  return example
+
+def extract_embedding_values(example_data: tf.train.Example) -> np.ndarray:
+  """
+  Extract the embeddings values contained in an Example object, extracted from a file
+  generated by the CXR foundation service.
+
+  Args:
+    example_data: The Example object to extract the values from
+
+  Returns:
+    The 1D float array of the embeddings values
+
+  Note: This is a convenience function for exploring/exporting. Do not use this in TF pipelines.
+
+  """
+  # Unpack nested proto and create np array from google.protobuf.pyext._message.RepeatedScalarContainer
+  try:
+    values = np.array(example_data.features.feature[constants.EMBEDDING_KEY].float_list.value, dtype='int32')
+    return values
+  except ValueError:
+    print(f"Input Example does not contain expected CXR Foundation embedding structure.")
+    raise
+
+def read_record_embeddings(filename: str) -> np.ndarray:
+  """
+  Read the embeddings values contained in a TFRecord embedding file, generated by the CXR
+  foundation service.
+
+  Args:
+    filename: The path of the .tfrecord file to read
+
+  Returns:
+    The 1D float array of the embeddings values
+
+  Note: This is a convenience function for exploring/exporting. Do not use this in TF pipelines.
+
+  """
+  return extract_embedding_values(read_record_example(filename))
+
+
+def parse_embedding_from_example(
+    serialized_example: bytes,
+) -> tf.Tensor:
+  """
+  Parses and extracts the embeddings values from a serialized tf.Example generated by the CXR foundation service.
+
+  Args:
+    serialized_example: The bytes of the tf.Example to parse.
+
+  Returns:
+    The 1D Tensor of float embeddings
+
+  """
+  features = {
+      constants.EMBEDDING_KEY:
+          tf.io.FixedLenFeature([constants.DEFAULT_EMBEDDINGS_SIZE],
+                                tf.float32,
+                                default_value=tf.constant(
+                                    0.0, shape=[constants.DEFAULT_EMBEDDINGS_SIZE]))
+  }
+  parsed_tensors = tf.io.parse_example(serialized_example, features=features)
+  return parsed_tensors[constants.EMBEDDING_KEY]
+
+
+def get_dataset(
+    filenames: Iterable[str],
+    labels: Iterable[int]
+) -> tf.data.Dataset:
+  """
+  Create a tf.data.Dataset from the specified tfrecord files and labels.
+
+  Args:
+    filenames: The set of .tfrecord file names.
+    labels: The corresponding label for each record.
+
+  Returns:
+    The Dataset, containing for each element: (embeddings, label)
+  """
+  ds_embeddings = tf.data.TFRecordDataset(filenames, num_parallel_reads=tf.data.AUTOTUNE).map(
+        parse_embedding_from_example
+   )
+  ds_labels = tf.data.Dataset.from_tensor_slices(labels)
+
+  return tf.data.Dataset.zip((ds_embeddings, ds_labels))
```

Comment:
Ref:
- https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_tensor_slices
- https://www.tensorflow.org/tutorials/load_data/tfrecord

I was/am actually rather confused about how the model training code doesn't reference label values at all. Then I inferred that they expect the elements of the Datasets to be: (value, label). Is this correct? Is there also an assumption of the weight being a possible third param?

Reply:
Yes, the model training code expects a tuple of (value, label, weight).